### PR TITLE
Add lumen ai enhancements

### PR DESCRIPTION
**Pull request by ahuang11**
Tons of changes, hopefully for the better:

Fixed:
- [x] Give the user high-level details about a particular dataset, i.e. list of columns and their types?! https://github.com/holoviz/lumen/pull/531/files#diff-d56253257f94b517eb742b1213dec9b7ccb3fafa7416be4044987ccd54ed3c83R234-R239
- [x] Chat about the data (same as above)
- [x] Ensure that it picks the correct agent, i.e. if we ask plot ensure that it will actually pick hvPlot (added a bit of short term memory https://github.com/holoviz/lumen/pull/531/files#diff-4699e4bd1882a5df97783ddd201518bcff9dc8ad2974a5c88196ffbc4c6cca38R173) 
- [x] User friendly error handling (“I’m sorry I did not understand”) (used pn.config.exception_handler here to catch all) https://github.com/holoviz/lumen/pull/531/files#diff-d56253257f94b517eb742b1213dec9b7ccb3fafa7416be4044987ccd54ed3c83R70
- [x] Add error handling so the model sees the error and/or Lumen validation output https://github.com/holoviz/lumen/pull/531/files#diff-d56253257f94b517eb742b1213dec9b7ccb3fafa7416be4044987ccd54ed3c83R650-R657
- [x] Need to figure out how it backs out of wrong pipelines and switch context https://github.com/holoviz/lumen/pull/531/files#diff-d56253257f94b517eb742b1213dec9b7ccb3fafa7416be4044987ccd54ed3c83R379-R382

Need help:
- [ ] Pick a particular table, perform some SQL transformation and then view the table and download it in various formats

Unable to reproduce:
- [ ] Scrolling is laggy (seems like this is with Llama + chat steps / did not implement here)
- [ ] Sometimes duplicates steps

### CODE REVIEW

File: lumen/ai/assistant.py [hunk 1]
```diff
@@ -33,12 +38,24 @@ def __init__(
         agents: list[Agent | Type[Agent]] | None = None,
         **params,
     ):
+        def download_messages():
+            def filter_by_reactions(messages):
+                return [
+                    message for message in messages if "favorite" in message.reactions
+                ]
+
+            messages = self.interface.serialize(filter_by=filter_by_reactions)
```
Comment (ahuang11) [inline lumen/ai/assistant.py:47]: This should use the custom serializer https://github.com/holoviz/lumen/pull/531/files#diff-4699e4bd1882a5df97783ddd201518bcff9dc8ad2974a5c88196ffbc4c6cca38R162-R169

So it only captures the YAML specs.

File: lumen/ai/assistant.py [hunk 2]
```diff
@@ -48,29 +65,47 @@ def __init__(
...
         )
         return prompt
 
     def _chat_invoke(self, contents: list | str, user: str, instance: ChatInterface):
+        print("-" * 50)
         return self.invoke(contents)
 
     def _choose_agent(self, messages: list | str, agents: list[Agent]):
-        agent_names = tuple(sagent.name for sagent in agents)
+        agent_names = tuple(sagent.name[:-5] for sagent in agents)
```
Comment (ahuang11) [inline lumen/ai/assistant.py:95]: Removing the numbers at the end like TableAgent01251

File: lumen/ai/assistant.py [hunk 3]
```diff
@@ -118,13 +154,36 @@ def _get_agent(self, messages: list | str):
...
+            if hasattr(obj, "object"):
+                return obj.object
+            elif hasattr(obj, "value"):
+                return obj.value
+            return str(obj)
+
         agent = self._get_agent(messages)
-        self._current_agent.object = f"## **Current Agent**: {agent.name}"
+        self._current_agent.object = f"## **Current Agent**: {agent.name[:-5]}"
+        messages = self.interface.serialize(custom_serializer=_custom_serializer)[-5:-1]
```
Comment (ahuang11) [inline lumen/ai/assistant.py:173]: Adds a bit of short term memory.

File: lumen/ai/llm.py [hunk 4]
```diff
@@ -54,16 +56,16 @@ def invoke(
         if system:
             messages = [{"role": "system", "content": system}] + messages
 
-        kwargs = dict(self._client_kwargs, **kwargs)
+        kwargs = dict(self._client_kwargs)
+        kwargs.update(input_kwargs)
```
Comment (ahuang11) [inline lumen/ai/llm.py:60]: I wanted to adjust GPT model per task.

**Comment by ahuang11** [comment]
In _models, we should standardize some model keys so it works across OpenAI, Mistral, etc because some tasks require larger models to reason and plan, while some tasks only need smaller models for quicker responses.

openai
```
_models = {
    "small": "gpt-3.5-turbo",
    "medium": "gpt-4",
    "large": "gpt-4"
}
```

mistralai
```
_models = {
    "small": "mistral-small",
    "medium": "mistral-medium",
    "large": "mistral-large"
}
```

File: lumen/ai/agents.py [hunk 5]
```diff
@@ -485,6 +502,7 @@ def _find_transform(
             messages,
             system=f"{system_prompt}\n{picker_prompt}",
             response_model=transform_model,
+            model="gpt-4",
```
Comment (philippjfr) [inline lumen/ai/agents.py:505]: Don't think this should be hardcoded here.

File: lumen/ai/agents.py [hunk 6]
```diff
@@ -576,7 +589,7 @@ class hvPlotAgent(LumenBaseAgent):
     """
 
     system_prompt = param.String(
-        default="Generate the plot the user requested. Note that x, y, by and groupby arguments may not reference the same columns."
+        default="Generate the plot the user requested. Note that x, y, by and groupby arguments may not reference the same columns. Be sure to add `download: csv`"
```
Comment (philippjfr) [inline lumen/ai/agents.py:592]: We can insert this after the fact.

**Comment by philippjfr** [comment]
Merging as-is for now.

**Pull request merged**